janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/optimize/strategy/AdjacentVertexHasUniquePropertyOptimizerStrategy.java

```diff
@@ -0,0 +1,145 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package org.janusgraph.graphdb.tinkerpop.optimize.strategy;
+
+import org.apache.tinkerpop.gremlin.process.computer.traversal.step.map.TraversalVertexProgramStep;
+import org.apache.tinkerpop.gremlin.process.traversal.*;
+import org.apache.tinkerpop.gremlin.process.traversal.dsl.graph.DefaultGraphTraversal;
+import org.apache.tinkerpop.gremlin.process.traversal.step.filter.FilterStep;
+import org.apache.tinkerpop.gremlin.process.traversal.step.filter.HasStep;
+import org.apache.tinkerpop.gremlin.process.traversal.step.map.GraphStep;
+import org.apache.tinkerpop.gremlin.process.traversal.step.map.IdStep;
+import org.apache.tinkerpop.gremlin.process.traversal.step.util.HasContainer;
+import org.apache.tinkerpop.gremlin.process.traversal.util.TraversalHelper;
+import org.apache.tinkerpop.gremlin.structure.Edge;
+import org.apache.tinkerpop.gremlin.structure.Graph;
+import org.apache.tinkerpop.gremlin.structure.Vertex;
+import org.janusgraph.core.Cardinality;
+import org.janusgraph.core.JanusGraphElement;
+import org.janusgraph.graphdb.database.StandardJanusGraph;
+import org.janusgraph.graphdb.query.JanusGraphPredicateUtils;
+import org.janusgraph.graphdb.query.QueryUtil;
+import org.janusgraph.graphdb.query.condition.*;
+import org.janusgraph.graphdb.query.index.IndexSelectionUtil;
+import org.janusgraph.graphdb.tinkerpop.optimize.JanusGraphTraversalUtil;
+import org.janusgraph.graphdb.transaction.StandardJanusGraphTx;
+import org.janusgraph.graphdb.types.CompositeIndexType;
+import org.janusgraph.graphdb.types.IndexType;
+
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static org.janusgraph.graphdb.types.system.ImplicitKey.ADJACENT_ID;
+
+/**
+ * @author Florian Grieskamp (Florian.Grieskamp@gdata.de)
+ */
+public class AdjacentVertexHasUniquePropertyOptimizerStrategy
+    extends AdjacentVertexOptimizerStrategy<HasStep<?>> {
+
+    private static final AdjacentVertexHasUniquePropertyOptimizerStrategy INSTANCE =
+        new AdjacentVertexHasUniquePropertyOptimizerStrategy();
+
+    private AdjacentVertexHasUniquePropertyOptimizerStrategy() {}
+
+    public static AdjacentVertexHasUniquePropertyOptimizerStrategy instance() { return INSTANCE; }
+
+    @Override
+    public Set<Class<? extends ProviderOptimizationStrategy>> applyPost() {
+        Set<Class<? extends ProviderOptimizationStrategy>> postStrategies = new HashSet<>();
+        postStrategies.add(JanusGraphStepStrategy.class);
+        postStrategies.add(JanusGraphLocalQueryOptimizerStrategy.class);
+        return postStrategies;
+    }
+
+    @Override
+    public void apply(final Traversal.Admin<?, ?> traversal) {
+        if (TraversalHelper.onGraphComputer(traversal)
+            || traversal.getStartStep() instanceof TraversalVertexProgramStep // not caught by TraversalHelper
```

Comment:
This looks interesting. How did you find this step? Did you know why it is not caught by TraversalHelper.onGraphComputer?

Reply:
This odd case occured to me while I was trying to fix failing tests that were atually failing because of the `StandardJanusGraph` vs `StandardJanusGraphTx` problem. While debugging I found out that `TraversalHelper.onGraphComputer()` does return false for traversals that start with `TraversalVertexProgramStep`s, which are obviously graph computer traversals. So I had a look into `TraversalHelper` (which is provided by TinkerPop, not JanusGraph) and what I found is that it only considers traversals as graph computer traversals if their _parent_ traversal begins with such a step. For the top layer traversal itself, it returns `false`.
What I also found out is that in current versions of TinkerPop, the implementation of this method has changed. However, I did not investigate whether or not this case is included in newer versions of TinkerPop.

When debugging, I saw that this change reduced the number of failing tests from 11 down to 2, so there must be some actual difference. However, all the 11 failing tests were fixed by introducing the `StandardJanusGraph`/`StandardJanusGraphTx` distinction, so this check isn't actually required. I still kept it in just to be sure, because it seemingly had a positive impact on the test runs earlier.